### ISSUE TITLE
Problem in reproducing fid score

### ISSUE THREAD

**Issue by howtowhy** [apple/ml-gmpi#22]
Thank you for amazing work again. 
 I have some trouble to reproducing fid score..

 1. I fail to download tfrecord and used 7GB public 256 dataset as a zip file. 
 (https://www.kaggle.com/datasets/denislukovnikov/ffhq256-images-only)
 2. I guess the ffhq256x256.zip file format is same as preprocessed one.
 3. I trained with this data and checked the training score.
 4. the fid goes up when I train..  from 20 to 25
 Could you give me some advice about this situation?
(I will try again with 1025 90G version with your advice..
Also,, if you give me some personal help please email me.. I will so appreciate your help.  howtowhy@gmail.com
I really want to refer your research but.. dataset problem is not easy for me... )

Thank you so much.

**Comment by Xiaoming-Zhao** [comment]
> I guess the ffhq256x256.zip file format is same as preprocessed one.

This one I am unsure as I have never tried the link you referred to before.

> I trained with this data and checked the training score.

This is weird. Do you mind sharing the hardware setup? E.g., the number of GPUs etc. For GAN's training, the batch size matters a lot. If possible, please use the same batch size as specified in the paper. 

> I really want to refer your research but.. dataset problem is not easy for me...

I am sorry and I definitely understand that this is a headache. BTW, I recently encountered a tool, i.e., [rclone](https://github.com/rclone/rclone), and it can interact with GDrive well. Please refer to the [documentation](https://rclone.org/drive/) for how to set it up. The only thing that needs some effort is to set up a Google Cloud API, which is not hard by following its [instruction](https://rclone.org/drive/#making-your-own-client-id).

Hope these help.

**Comment by Xiaoming-Zhao** [comment]
Hi @parkjh688, I need several more information if possible.

> In the original paper, the model was trained with a batch size of 64, so I also tried that

How many GPUs did you use to train GMPI? The reason I am asking is that the `batch_size` specified in the [`curriculums.py`](https://github.com/apple/ml-gmpi/blob/672294b56f97bd621a87a7a33a39964abb30c0bc/gmpi/curriculums.py) is for batch size per GPU. The batch size of 64 stated in the paper comes from `8 (per GPU) x 8 (#GPUs) = 64`.

> Previously, when I trained with a batch size of 8, the FID score was around 20 and remained stable until the end of 5000 steps.

What is the dataset you used for this training and how many GPUs were you using? Essentially, I want to know your actual batch size.

> I'm wondering if using ffhq256x256 data from Kaggle instead of real 256-sized data could be causing overfitting.

One caveat I could see is that the Kaggle dataset uses a different resizing method from the one the pre-trained StyleGAN2 uses. Specifically:
a. The Kaggle one uses `bicubic` as specified on the webpage.
b. The StyleGAN2 is trained on images downscaled with `lanczos` (see [here](https://github.com/NVlabs/stylegan2-ada-pytorch/blob/6f160b3d22b8b178ebe533a50d4d5e63aedba21d/dataset_tool.py#L309)).

So maybe you want to process the dataset following [the official instruction](https://github.com/NVlabs/stylegan2-ada-pytorch/blob/6f160b3d22b8b178ebe533a50d4d5e63aedba21d/README.md#preparing-datasets) to have a double check.

>  I trained my model with the following configuration: 256: {'batch_size': 64, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 16, 'gen_lr': 0.002, 'disc_lr': 0.002}

I noticed that you have `batch_split` of 16. This means that you will split the 64 images into 16 mini-batches and accumulate gradients 16 times. Theoretically, this should be fine. However, I would recommend using only one single forward pass if possible to avoid any hidden issues.

Hope these help.

**Comment by parkjh688** [comment]
> The reason I am asking is that the batch_size specified in the [curriculums.py](https://github.com/apple/ml-gmpi/blob/672294b56f97bd621a87a7a33a39964abb30c0bc/gmpi/curriculums.py) is for batch size per GPU. The batch size of 64 stated in the paper comes from 8 (per GPU) x 8 (#GPUs) = 64.

Oh, I didn't know that. I used 6 GPUs and my configuration in curriculums.py was 256: {'batch_size': 64, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 16, 'gen_lr': 0.002, 'disc_lr': 0.002} this. Then if I want to train the model with 64 batch size I have to train it with 4 GPUs and the configuration will be 256: {'batch_size': 16, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 1, 'gen_lr': 0.002, 'disc_lr': 0.002} like this. 16 (per GPU) x 4 (#GPUs) = 64

> What is the dataset you used for this training? And how many GPUs were you using?

I use this [kaggle dataset.](https://www.kaggle.com/datasets/denislukovnikov/ffhq256-images-only).

Thanks!

**Comment by Xiaoming-Zhao** [comment]
Got you. So the Kaggle dataset is indeed able to reproduce FID based on your previous statement
> Previously, when I trained with a batch size of 8, the FID score was around 20 and remained stable until the end of 5000 steps.

Then I would recommend reducing the `batch_split` to see whether this is the culprit for the weird FID curve you showed before. As I mentioned before, theoretically, having `batch_split = 16` should be fine but I am not sure whether there will be some hidden issues there.

Hope this helps.

**Comment by howtowhy** [comment]
Hello, thank you for your detailed help.
I downloaded the 1024 image and preprocessing to 256 with script.
And I used following option with 8 GPU

"res_dict": {
256: {'batch_size': 64, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 16, 'gen_lr': 0.002, 'disc_lr': 0.002},
512: {'batch_size': 4, 'num_steps': 32, 'img_size': 512, 'tex_size': 512, 'batch_split': 1, 'gen_lr': 0.002, 'disc_lr': 0.002},
1024: {'batch_size': 4, 'num_steps': 32, 'img_size': 1024, 'tex_size': 1024, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
},

```
"res_dict_learnable_param": {
    256: {'batch_size': 64, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 16, 'gen_lr': 0.002, 'disc_lr': 0.002},
    512: {'batch_size': 4, 'num_steps': 32, 'img_size': 512, 'tex_size': 512, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
    1024: {'batch_size': 4, 'num_steps': 32, 'img_size': 1024, 'tex_size': 1024, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
},

```

But the fid goes up and the result was like this.
Could you advice for this situation? 

![Uploading image.png…]()

**Comment by howtowhy** [comment]
Hello! Thank you for your kind help.
I run the script with 8 GPU and batchsize you suggested.

```
"res_dict": {
        256: {'batch_size': 8, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 1, 'gen_lr': 0.002, 'disc_lr': 0.002},
        512: {'batch_size': 4, 'num_steps': 32, 'img_size': 512, 'tex_size': 512, 'batch_split': 1, 'gen_lr': 0.002, 'disc_lr': 0.002},
        1024: {'batch_size': 4, 'num_steps': 32, 'img_size': 1024, 'tex_size': 1024, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
    },

    "res_dict_learnable_param": {
        256: {'batch_size': 8, 'num_steps': 32, 'img_size': 256, 'tex_size': 256, 'batch_split': 1, 'gen_lr': 0.002, 'disc_lr': 0.002},
        512: {'batch_size': 4, 'num_steps': 32, 'img_size': 512, 'tex_size': 512, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
        1024: {'batch_size': 4, 'num_steps': 32, 'img_size': 1024, 'tex_size': 1024, 'batch_split': 2, 'gen_lr': 0.002, 'disc_lr': 0.002},
    },
```

The fid score with 256 was 18.92-21.52. (with one peak pertubation)
But the paper suggest its fid to 11.4
Is there any problem I missed?
Thank you for your quick help so much.

<img width="386" alt="image" src="https://user-images.githubusercontent.com/10994112/233585886-0a815132-c5ae-4da5-9afd-4874caf6c610.png">

**Comment by Xiaoming-Zhao** [comment]
This curve looks reasonable to me. I am not sure about the peak but I guess it may due to some randomness.

Regarding the FID: FID score largely depends on the number of images used to compute the score. The more images you use, the large probability you will obtain a lower score.

However, FID with plenty of images is costly to compute. Therefore, during training, we use a small number of images to get a sense of the FID trend:
- `#real_images = 8000`: https://github.com/apple/ml-gmpi/blob/672294b56f97bd621a87a7a33a39964abb30c0bc/gmpi/train.py#L1013
- `#fake_images=2048`: https://github.com/apple/ml-gmpi/blob/672294b56f97bd621a87a7a33a39964abb30c0bc/gmpi/train.py#L1025

During the full evaluation, we use 50k fake and real images as stated in the paper: https://github.com/apple/ml-gmpi/blob/672294b56f97bd621a87a7a33a39964abb30c0bc/gmpi/eval/eval.sh#L17

Hope this resolves your confusion.
